src/DotNetty.Common/ThreadLocalPool.cs

```diff
@@ -41,21 +50,307 @@ public bool Release<T>(T value)
             }
         }
 
-        internal sealed class Stack : Stack<Handle>
+        internal sealed class WeakOrderQueue
         {
-            public readonly ThreadLocalPool Owner;
+            const int LINK_CAPACITY = 16;
+            class Link
+            {                
+                internal int readIndex;
+                internal volatile int writeIndex;
```

Comment:
pls switch to Volatile.Read/Write methods. see http://blog.alexrp.com/2014/03/30/dot-net-atomics-and-memory-model-semantics/ for details